# Citations

If you use plexus in your research, please cite the tools and resources it depends on.

## Primer Design

**Primer3**
Untergasser, A., Cutcutache, I., Koressaar, T., Ye, J., Faircloth, B.C., Remm, M., & Rozen, S.G. (2012). Primer3 â€” new capabilities and interfaces. *Nucleic Acids Research*, 40(15), e115. https://doi.org/10.1093/nar/gks596

Koressaar, T., & Remm, M. (2007). Enhancements and modifications of primer design program Primer3. *Bioinformatics*, 23(10), 1289-1291. https://doi.org/10.1093/bioinformatics/btm091

## Thermodynamic Models

**SantaLucia (1998) Nearest-Neighbor Model**
SantaLucia, J. (1998). A unified view of polymer, dumbbell, and oligonucleotide DNA nearest-neighbor thermodynamics. *Proceedings of the National Academy of Sciences*, 95(4), 1460-1465. https://doi.org/10.1073/pnas.95.4.1460

## Primer-Dimer Prediction

**Johnston et al. (2019)**
Johnston, A.D., Lu, J., Ru, K., Korbie, D., & Trau, M. (2019). PrimerROC: accurate condition-independent dimer prediction using nearest neighbor base stacking. *Scientific Reports*, 9, 9218. https://doi.org/10.1038/s41598-019-45599-2

## Sequence Alignment and BLAST

**NCBI BLAST**
Altschul, S.F., Gish, W., Miller, W., Myers, E.W., & Lipman, D.J. (1990). Basic local alignment search tool. *Journal of Molecular Biology*, 215(3), 403-410. https://doi.org/10.1016/S0022-2836(05)80360-2

**Biopython**
Cock, P.J., Antao, T., Chang, J.T., Chapman, B.A., Cox, C.J., Dalke, A., Friedberg, I., Hamelryck, T., Kauff, F., Wilczynski, B., & de Hoon, M.J. (2009). Biopython: freely available Python tools for computational molecular biology and bioinformatics. *Bioinformatics*, 25(11), 1422-1423. https://doi.org/10.1093/bioinformatics/btp163

## Genomic Data Access

**HTSlib / SAMtools**
Danecek, P., Bonfield, J.K., Liddle, J., Marshall, J., Ober, V., Pollard, M.O., Whitwham, A., Keane, T., McCarthy, S.A., Davies, R.M., & Li, H. (2021). Twelve years of SAMtools and BCFtools. *GigaScience*, 10(2), giab008. https://doi.org/10.1093/gigascience/giab008

## Population Variant Data

**gnomAD**
Karczewski, K.J., Francioli, L.C., Tiao, G., Cummings, B.B., Alfoldi, J., Wang, Q., Collins, R.L., Laricchia, K.M., Ganna, A., Birnbaum, D.P., et al. (2020). The mutational constraint spectrum quantified from variation in 141,456 humans. *Nature*, 581(7809), 434-443. https://doi.org/10.1038/s41586-020-2308-7
